From: Michal Kubecek <mkubecek@suse.cz>
Date: Fri, 5 Sep 2014 11:25:54 +0200
Subject: net: fix checksumming features handling in output path
Patch-mainline: Queued in subsystem maintainer repository
Git-repo: git://git.kernel.org/pub/scm/linux/kernel/git/davem/net.git
Git-commit: db115037bb57cdfe97078b13da762213f7980e81
References: bnc#891259

Since checksumming features handling was fixed for vlan devices
on top of a bond (bnc#872634), bonding device can have IP_CSUM
in features and HW_CSUM in vlan_features. In such case,
netif_skb_features() called from dev_hard_start_xmit() for the
bond will do a bitwise and and detect no checksumming support,
resulting in software checksum calculation and CHECKSUM_NONE.
When the packet is passed to the slave, this may trigger the
WARN in skb_gso_segment().

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 net/core/dev.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 0aec89f..2ae2d12 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -2075,13 +2075,19 @@ u32 netif_skb_features(struct sk_buff *skb)
 		return harmonize_features(skb, protocol, features);
 	}
 
-	features &= (skb->dev->vlan_features | NETIF_F_HW_VLAN_TX);
+	features = netdev_intersect_features(features,
+					     skb->dev->vlan_features |
+					     NETIF_F_HW_VLAN_TX);
 
 	if (protocol != htons(ETH_P_8021Q)) {
 		return harmonize_features(skb, protocol, features);
 	} else {
-		features &= NETIF_F_SG | NETIF_F_HIGHDMA | NETIF_F_FRAGLIST |
-				NETIF_F_GEN_CSUM | NETIF_F_HW_VLAN_TX;
+		features = netdev_intersect_features(features,
+						     NETIF_F_SG |
+						     NETIF_F_HIGHDMA |
+						     NETIF_F_FRAGLIST |
+						     NETIF_F_GEN_CSUM |
+						     NETIF_F_HW_VLAN_TX);
 		return harmonize_features(skb, protocol, features);
 	}
 }
-- 
1.8.4.5

